requires unique names for running sites

tannerjfco opened the pull request:
## The Problem:
It is currently possible to start a site that has the same name of a site already running. This is undesirable, as it can apparently allow the new site's containers to replace the running site's containers without warning (per OP #136). 

## The Fix:
This introduces a check of containers in `Start()` to see if any match the name we are about to start and are running. If there's a match, an error will return and the site won't be started.

## The Test:
- Have 2 sites to test
- Set site 1's name (by config.yaml or ddev config) to "testsite"
- `ddev start` for site 1 should succeed
- Set site 2's name to "testsite" (same as site 1)
- `ddev start` for site 2 should immediately fail with error.

## Automation Overview:
<!-- Please provide an overview of tests introduced by this PR, or an explanation for why no tests are needed. -->
A test case has been added to `TestLocalStart()` that attempts to start the first test site again, triggering this error.

## Related Issue Link(s):
#136 OP
## Release/Deployment notes:
<!-- Does this affect anything else, or are there ramifications for other code? Does anything have to be done on deployment? -->

rfay commented on pkg/plugins/platform/local.go:
```diff
@@ -258,6 +258,15 @@ func (l *LocalApp) Start() error {
 	composePath := l.DockerComposeYAMLPath()
 	l.DockerEnv()
 
+	existing, err := l.FindContainerByType("web")
```
I don't think this actually resolves the key problem here because it assumes that the competing site is running, which might or might not be the case. I understand that we're storing all our state info in container labels, which prevents us from a robust approach here. 

I'm not sure what the correct approach is here, but this doesn't fully solve https://github.com/drud/ddev/issues/136 - and of course not mentioned there are the facts that `ddev rm` can destroy a site from another directory, etc.  I imagine we have to have a conversation about how much of it is worth solving, and whether we can do it with our current level of config (mostly stashed on running containers, and nothing global.)

tannerjfco replied:
You're right, this doesn't fully resolve the issues identified in #136. I do think this is the right approach in terms of the check however. It should be possible for a user to stop site 1 version A, make a copy site 1 version B, and start version B. 

I still don't think we should auto-adjust the site name, but I do think we can change the container name pattern to append a unique hash at the end. With Brad's list rework, I don't think we should be depending on the container name pattern at this point, and should instead be leveraging labels, so I think this could be a reasonable change at this stage.

beeradb replied:
`FindContainerByType()` will actually return stopped containers as well.

beeradb commented on pkg/plugins/platform/local.go:
```diff
@@ -258,6 +258,15 @@ func (l *LocalApp) Start() error {
+		return err
+	}
+
+	if existing.State == "running" {
```
There is a `SiteStatus()` method coming in the list rework. It would probably be wise to use that abstraction once it lands over accessing state directly.

beeradb replied:
This code can all be removed now since it all gets checked in init(), right?

tannerjfco commented:
I've updated this PR to better address the issues surrounding #136. In addition to the check on ddev start for already running site, I've added a check for all commands that involve interaction with docker. Init will now check for any existing containers, and if they exist, compares the approot from container label against the approot path the command is trying to run against. This should prevent the scenarios @rfay described from occurring.

beeradb commented on pkg/plugins/platform/local.go:
```diff
@@ -49,6 +49,14 @@ func (l *LocalApp) Init(basePath string) error {
 
+	web, err := l.FindContainerByType("web")
+	if err == nil {
+		containerApproot := web.Labels["com.ddev.approot"]
```
Curious if we should handle the unset/empty value ("") here. What should the behavior be? should all commands work? none?

beeradb commented:
Other than the feedback above this worked as expected. Thanks!

rickmanelius commented on pkg/plugins/platform/local_test.go:
```diff
@@ -131,6 +131,24 @@ func TestLocalStart(t *testing.T) {
+	assert.NoError(err)
+
+	err = app.Start()
+	assert.EqualError(err, fmt.Sprintf("a site with the name %s is already running", TestSites[0].Name))
```
IMHO, this tells the user that there is a problem without necessarily pointing them to a resolution. For example, I would have to remember to run `ddev config` to change the site name.

tannerjfco replied:
This check was intended to prevent a site that is already running from being started again, but I hadn't considered the use case for re-running start on an existing site. It is now removed.

rickmanelius replied:
I might not be understanding fully. Just to make sure, let me rephase my first comment a different way. If I had two different "wordpress" repos and I started (1) and was about to start (2), this would be a proper warning. The issue from an end-user would be "ok, so how do I fix that?", which is why I was suggesting a hint about how to resolve by re-running ddev config.

tannerjfco replied:
This specific message would have only been surfaced if a user tried to run start on the same exact site again.

The message you would see in the scenario you describe is here: https://github.com/drud/ddev/pull/169/files#diff-ef6770d9874ceccc0d1f5c859385f6feR56 That should indicate to the user the state of the container found and the location the found container was started from.

rfay commented:
List stuff was merged so this could rebase now and use the container status function.

tannerjfco commented:
The status check was part of the check that was removed from start, so this can be good to go once @beeradb approves.

rfay commented:
Your problems with this are now resolved, right @beeradb ?
